qsimcirq/qsim_simulator.py

```diff
@@ -323,30 +323,35 @@ def _sample_measure_results(
                 cirq.MeasurementGate
             )
         ]
-        measured_qubits: List[cirq.Qid] = []
-        bounds: Dict[str, Tuple] = {}
+        num_qubits_by_key: Dict[str, int] = {}
+        bounds: Dict[str, Tuple[int, int]] = {}
         meas_ops: Dict[str, List[cirq.GateOperation]] = {}
         current_index = 0
         for op in measurement_ops:
             gate = op.gate
             key = cirq.measurement_key_name(gate)
             meas_ops.setdefault(key, [])
             meas_ops[key].append(op)
-            if key in bounds:
-                raise ValueError(f"Duplicate MeasurementGate with key {key}")
-            bounds[key] = (current_index, current_index + len(op.qubits))
-            measured_qubits.extend(op.qubits)
-            current_index += len(op.qubits)
+            n = len(op.qubits)
+            if key in num_qubits_by_key:
+                if n != num_qubits_by_key[key]:
+                    raise ValueError(
+                        f"repeated key {key!r} with different numbers of qubits: "
+                        f"{num_qubits_by_key[key]} != {n}"
+                    )
+            else:
+                num_qubits_by_key[key] = n
+            if key not in bounds:
+                bounds[key] = (current_index, current_index + n)
```

Comment:
`bounds` is used at the bottom of this function to determine how to map measurement results from qsim's raw result output to Cirq's keyed results dictionary. In order for this to work, we'll need bounds stored for each instance of the given measurement key.